### PR TITLE
ARROW-9483: [C++] Reorganize testing headers

### PR DESCRIPTION
**Pull request by pitrou**
* Include less headers transitively from `gtest_util.h`.
* Remove `gtest_common.h`, use `RandomArrayGenerator` instead.
* Split `visitor_inline.h` into several topic-specific headers

### CODE REVIEW

**Comment by lidavidm** [comment]
Tangential question, do we still use IWYU for this sort of thing?

**Comment by pitrou** [comment]
Well, as far as I'm concerned, I never managed to get IWYU to tell anything useful :-)

File: cpp/src/arrow/visit_scalar_inline.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  switch (scalar.type->id()) {
+    ARROW_GENERATE_FOR_ALL_TYPES(SCALAR_VISIT_INLINE);
+    default:
+      break;
+  }
+  return Status::NotImplemented("Scalar visitor for type not implemented ",
+                                scalar.type->ToString());
+}
+
+#undef TYPE_VISIT_INLINE
```
Comment (lidavidm) [inline cpp/src/arrow/visit_scalar_inline.h:46]: SCALAR_VISIT_INLINE?

**Review by lidavidm — APPROVED** [review]

File: cpp/src/arrow/visit_scalar_inline.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/visit_scalar_inline.h:46]: Oops, certainly.

**Review by lidavidm — APPROVED** [review]

**Pull request closed without merging**